### PR TITLE
[JSX] Fixed check for select elements with multiple attribute 

### PR DESCRIPTION
**Pull request by llevitis**
This PR modifies the FilterForm & StaticDataTable components to handle SelectElements where the multiple attribute is selected. Previously, only strings and numbers were handled in the StaticDataTable, and now the 'object' type is being checked for. Additionally, queryString.js now handles an array of values being passed for a single filter element.

### CODE REVIEW

**Comment by PapillonMcGill** [comment]
Do we allow multiple return statement in a single function? Or I am old school?

File: jsx/StaticDataTable.js [hunk 1]
```diff
@@ -297,6 +297,20 @@ var StaticDataTable = React.createClass({
...
+    // Handle array inputs for multiselects
+    if (typeof filterData === 'object') {
+      var match = false;
+      for (let i = 0; i < filterData.length; i += 1) {
+        var searchKey = filterData[i].toLowerCase();
+        var searchString = data.toLowerCase();
+
+        match = (searchString.indexOf(searchKey) > -1);
+        if (match) {
+          return true;
```
Comment (HenriRabalais) [inline jsx/StaticDataTable.js]: Perhaps a 'break' is more appropriate here since the 'return match' on line 312 will be returning true under this condition as well - that is if we are trying to avoid multiple return statements. I'm not sure what the convention is on that one.

Comment (llevitis) [reply jsx/StaticDataTable.js]: @PapillonMcGill & @HenriRabalais - I followed the convention that had been in place for the existing checks (for number & string inputs). I wouldn't say that the function is returning multiple things, it just returns 'true' if a certain condition is met and if not, it returns the output of another check.

Comment (PapillonMcGill) [reply jsx/StaticDataTable.js]: @llevitis The multiple return statement will of course return only 1 thing (true or false), but previously, it what considered a bad practice to have multiple return statement inside a function, you would set a variable ($result = true) mid function as needed with possible break statement and a the end, return the variable value.  
Maybe things have change but I still only use 1 return statement per function as it make it clear what is returned, just look at the last line.

**Comment by ridz1208** [comment]
@llevitis please put a branch label on your PR

File: htdocs/js/util/queryString.js [hunk 2]
```diff
@@ -50,15 +50,6 @@ var QueryString = {
     // Deep copy of object representation of query
     var queryStringObj = JSON.parse(JSON.stringify(currentQuery));
 
-    // Make sure that both key and value are of string type
-    if (typeof fieldName !== "string" || typeof fieldValue !== "string") {
```
Comment (davidblader) [inline htdocs/js/util/queryString.js]: since the only expected change in type is for fieldValue, the check on fieldName is still valid. 
i would either:
a) remove `|| typeof fieldValue !== "string"` 
or
b) change it to `|| (typeof fieldValue !== "string" && typeof fieldValue !== "object")` 
and then update the error message accordingly. 

the documentation in the comment above should also be amended to suggest that fieldValue can be either string or object. 
same comment applies to the changes in FilterForm.js

File: jsx/StaticDataTable.js [hunk 3]
```diff
@@ -297,6 +297,19 @@ var StaticDataTable = React.createClass({
       return (searchString.indexOf(searchKey) > -1);
     }
 
+    // Handle array inputs for multiselects
+    if (typeof filterData === 'object') {
+      var match = false;
+      for (let i = 0; i < filterData.length; i += 1) {
+        var searchKey2 = filterData[i].toLowerCase();
```
Comment (davidblader) [inline jsx/StaticDataTable.js]: definitely not a big deal but just curious as to why you named this `searchKey2` instead of just `searchKey`

Comment (llevitis) [reply jsx/StaticDataTable.js]: it was failing Travis since the var searchKey had already been defined before..I could create a more global variable called 'searchKey' instead

Comment (xlecours) [reply jsx/StaticDataTable.js]: Or just change the `var` for `let`

File: jsx/StaticDataTable.js (hunk 1 above)
Comment (xlecours) [reply jsx/StaticDataTable.js]: @llevitis please use `let` or `const` to declare variables that value can or can't change

Also, what you did is just fine (you followed the existing convention to keep coherence in this function code). However, as @HenriRabalais and @PapillonMcGill are saying, it is makes it hard to follow the flow of the function when there is multiple return statement in it. With the exception of some obvious checks/assertions at the start of the function it is preferable that a return variable is declared at the top of the function and its value assignment (optionally) occur later. 
Now, if you don't mind, you could: 1. declare a return value initially set to false. 2. use the `||` operator to get the value of `this.props.Filter`. 3. use a switch statement on `typeof filter.value` then put the comparison code in `case "number":` `case "string"` and `case "object"`. 4. a single return at the bottom

File: jsx/StaticDataTable.js [hunk 4]
```diff
@@ -282,22 +285,35 @@ var StaticDataTable = React.createClass({
...
       }
 
-      return (searchString.indexOf(searchKey) > -1);
+      result = (searchString.indexOf(searchKey) > -1);
     }
 
-    return false;
+    // Handle array inputs for multiselects
+    if (typeof filterData === 'object') {
+      var match = false;
```
Comment (xlecours) [inline jsx/StaticDataTable.js]: `let`

File: jsx/StaticDataTable.js [hunk 5]
```diff
@@ -282,22 +285,35 @@ var StaticDataTable = React.createClass({
...
+        result = (searchString === searchKey);
       }
 
-      return (searchString.indexOf(searchKey) > -1);
+      result = (searchString.indexOf(searchKey) > -1);
     }
 
-    return false;
+    // Handle array inputs for multiselects
+    if (typeof filterData === 'object') {
```
Comment (davidblader) [inline jsx/StaticDataTable.js]: issue for when you unselect all options, everything gets filtered out

Comment (davidblader) [reply jsx/StaticDataTable.js]: after unselecting:
image
DataDictIndex state:
image

Comment (HenriRabalais) [reply jsx/StaticDataTable.js]: There's an inconsistency with the use of quotes when checking `typeof`. I suggest changing the boolean expression checks on lines `54`, `150`, and `151` to single quotes as well since they are more appropriate for those cirumstances.

**Comment by ridz1208** [comment]
@xlecours please dismiss and/or re-review

**Comment by zaliqarosli** [comment]
trying to test this on Firefox and there seems to be no changes to my instruments select element

**Comment by ridz1208** [comment]
@zaliqarosli it needs to function on firefox yes
@llevitis please look into the firefox issue

@HenriRabalais were your comments addressed ?

**Comment by davidblader** [comment]
@ridz1208 i did not encounter any issues testing on firefox

**Comment by llevitis** [comment]
@davidblader - I think the issue was that @zaliqarosli made the comment before I properly explained how to test it. thank u for re-testing!

**Comment by zaliqarosli** [comment]
@ridz1208 @llevitis @davidblader yep! i never properly tested it so ignore my comment!

**Comment by driusan** [comment]
@llevitis conflicts
